### PR TITLE
Fixing the last deviations from sentencepiece indicated by test-tokenizer-1

### PR DESCRIPTION
**Pull request by goerch**
The [last deviations](https://github.com/ggerganov/llama.cpp/pull/3096) are fixed now, too.

We don't seem to need Unicode normalization immediately.

### CODE REVIEW

File: common/common.cpp [hunk 1]
```diff
@@ -804,10 +804,10 @@ std::vector<llama_token> llama_tokenize(
     // upper limit for the number of tokens
     int n_tokens = text.length() + add_bos;
     std::vector<llama_token> result(n_tokens);
-    n_tokens = llama_tokenize(ctx, text.c_str(), result.data(), result.size(), add_bos);
+    n_tokens = llama_tokenize(ctx, text.c_str(), text.length(), result.data(), result.size(), add_bos);
```
Comment (Cebtenzzre) [inline common/common.cpp:807]: Why is this needed?

Comment (goerch) [reply common/common.cpp:807]: Otherwise I get these 

```
main : error: token 3 detokenizes to ><(1) but tokenization of this detokenizes to ><(0)
main : error: codepoint 0 detokenizes to ><(0) instead of ><(1)
```

errors from `test-tokenizer-1`. C strings do not differentiate "\0" from "\0\0".

Comment (Cebtenzzre) [reply common/common.cpp:807]: But should that be possible to run into in practice? For UTF-8 text, the NUL byte should only appear at the end of a string. Who would be trying to pass binary data to the tokenizer?

Comment (goerch) [reply common/common.cpp:807]: > But should that be possible to run into in practice? 

I'm not sure, but these test cases work for the `sentencepiece` tokenizer.

**Review by ggerganov — APPROVED** [review]

**Review by cebtenzzre — CHANGES_REQUESTED** [review]
I would like a more complete understanding of why the API change is needed. The fact that sentencepiece accepts it isn't good enough - does the `text` parameter represent a UTF-8 string, or something else?

**Comment by goerch** [comment]
> I would like a more complete understanding of why the API change is needed. The fact that sentencepiece accepts it isn't good enough - does the `text` parameter represent a UTF-8 string, or something else?

As far as I understand it, '0x00' is a valid Unicode point which converts to '0x00'.

File: common/common.cpp (hunk 1 above)
Comment (goerch) [reply common/common.cpp:807]: > For UTF-8 text, the NUL byte should only appear at the end of a string.

Not sure if this is correct.

**Comment by goerch** [comment]
> > As far as I understand it, `0x00` is a valid Unicode point which converts to `0x00` in UTF8.
> 
> Yes, but it is a non-printable control character. As far as [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_403) is concerned, text files must not contain NUL bytes - therefore, text should not contain NUL bytes.

The small print says: 'Although POSIX.1-2017 does not distinguish between text files and binary files (see the ISO C standard)...'

**Comment by Cebtenzzre** [comment]
If the POSIX description isn't sufficient, what about MIME types?
```console
$ printf 'hello\n' >text.bin
$ printf 'hello\0\n' >not_text.bin
$ file -i text.bin
text.bin: text/plain; charset=us-ascii
$ file -i not_text.bin
not_text.bin: application/octet-stream; charset=binary
```

**Comment by goerch** [comment]
@Cebtenzzre : I believe that the NUL character is improbable and badly supported by software around. But this PR is aiming to improve `sentencepiece` compatibility for a specific test case. So either we work with this PR or change the test case. How do you propose I should change the test case? Ignore token 3 and code point 0? Then I can simply retract this change.

**Comment by slaren** [comment]
I am not sure why POSIX is relevant here. Is there any reason to believe that the tokenizer needs to respect any part of the POSIX spec? I think this is simpler that, if the sentencepiece tokenizer can encode NUL characters, but our implementation can't, then it is a bug in our implementation and it should be fixed. Supporting non-NUL terminated strings seems the natural way to do it.

**Review by cebtenzzre — APPROVED** [review]
Okay, I'll defer to your judgement in this case.

File: common/common.cpp (hunk 1 above)
Comment (Cebtenzzre) [inline common/common.cpp:807]:
```suggestion
    n_tokens = llama_tokenize(ctx, text.data(), text.length(), result.data(), result.size(), add_bos);
```

Comment (goerch) [reply common/common.cpp:807]: Thanks! Will change.

Comment (goerch) [reply common/common.cpp]: Stupid me didn't simply accept your changes, sorry!

Comment (Cebtenzzre) [reply common/common.cpp]: It's fine, it would have only applied to the first case anyway.

**Comment by alitariq4589** [comment]
Can one of the admins verify this patch?
